File: day4/main.lua
package.path = "../util/?.lua;" .. package.path

local filereader = require "filereader"

local input = filereader.lines_from('challenge_input.txt')

function split(s, delimiter)
     local result = {};
    for match in (s..delimiter):gmatch("(.-)"..delimiter) do
        table.insert(result, match);
    end
    return result;
end

local score = 0
for index,sections in pairs(input) do
  local splitPairs = split(sections, ',')
  local splitPair1 = split(splitPairs[1], '-')
  local splitPair2 = split(splitPairs[2], '-')

  if ((splitPair1[1] <= splitPair2[1] and splitPair1[2] <= splitPair2[2]) or (splitPair1[1] >= splitPair2[1] and splitPair1[2] >= splitPair2[2])) then
    score = score +1
  end

end

print(score)

File: day4/main.py
def readInput():
    with open('challenge_input.txt') as f:
        return f.readlines()

def isFullyOverlapping(splitPair1, splitPair2):
    if int(splitPair1[0]) >= int(splitPair2[0]) and int(splitPair1[1]) <= int(splitPair2[1]) or int(splitPair1[0]) <= int(splitPair2[0]) and int(splitPair1[1]) >= int(splitPair2[1]):
        return True
    else:
        return False

def isPartlyOverlapping(splitPair1, splitPair2):
    set1 = set(range(int(splitPair1[0]),  int(splitPair1[1])+1))
    set2 = set(range(int(splitPair2[0]), int(splitPair2[1])+1))
    intersection = set1.intersection(set2)

    if (intersection):
        return True
    else:
        return False

def main():
    # Part One
    input = readInput()
    fullyScore = 0
    partlyScore = 0
    # '31-69,35-55\n'
    for pairs in input:
        pairs = pairs.strip()
        splitPairs = pairs.split(',')
        splitPair1 = splitPairs[0].split('-')
        splitPair2 = splitPairs[1].split('-')

        fullyOverlapping = isFullyOverlapping(splitPair1, splitPair2)
        partlyOverlapping = isPartlyOverlapping(splitPair1, splitPair2)

        if fullyOverlapping:
            fullyScore += 1

        if partlyOverlapping:
            partlyScore += 1




    print('Pairs fully overlapping: ', fullyScore)
    print('Pairs partly overlapping: ', partlyScore)

if __name__ == "__main__":
    main()
